src/libraries/System.Private.CoreLib/src/System/IO/RandomAccess.Windows.cs

```diff
@@ -421,82 +424,190 @@ internal static void WriteGatherAtOffset(SafeFileHandle handle, IReadOnlyList<Re
 
-        private static async ValueTask<long> ReadScatterAtOffsetSingleSyscallAsync(SafeFileHandle handle, IReadOnlyList<Memory<byte>> buffers, long fileOffset, int totalBytes, CancellationToken cancellationToken)
+        // From the same source:
+        // "Each buffer must be at least the size of a system memory page and must be aligned on a system
+        // memory page size boundary. The system reads/writes one system memory page of data into/from each buffer."
+        // This method returns true if the buffers can be used by
+        // the Windows scatter/gather API, which happens when they are:
+        // 1. aligned at page size boundaries
+        // 2. exactly one page long each (our own requirement to prevent partial reads)
+        // 3. not bigger than 2^32 - 1 in total
+        // This function is also responsible for pinning the buffers if they
+        // are suitable and they must be unpinned after the I/O operation completes.
+        // The total size of the buffers is also returned.
+        private static bool TryPrepareScatterGatherBuffers<T, THandler>(IReadOnlyList<T> buffers,
+            THandler handler, [NotNullWhen(true)] out MemoryHandle[]? pinnedBuffers, out int totalBytes)
+            where THandler: struct, IMemoryHandler<T>
         {
+            int pageSize = Environment.SystemPageSize;
+            Debug.Assert(BitOperations.IsPow2(pageSize), "Page size is not a power of two.");
+            // We take advantage of the fact that the page size is
+            // a power of two to avoid an expensive modulo operation.
+            long alignedAtPageSizeMask = pageSize - 1;
             int buffersCount = buffers.Count;
-            if (buffersCount == 1)
+            pinnedBuffers = new MemoryHandle[buffersCount];
+
+            try
             {
-                // we have to await it because we can't cast a VT<int> to VT<long>
-                return await ReadAtOffsetAsync(handle, buffers[0], fileOffset, cancellationToken).ConfigureAwait(false);
+                long totalBytes64 = 0;
+                for (int i = 0; i < buffersCount; i++)
+                {
+                    T buffer = buffers[i];
+                    int length = handler.GetLength(in buffer);
+                    totalBytes64 += length;
+                    if (length != pageSize || totalBytes64 > int.MaxValue)
+                    {
+                        goto Failure;
+                    }
+
+                    MemoryHandle handle = pinnedBuffers[i] = handler.Pin(in buffer);
+                    unsafe
+                    {
+                        if (((long)handle.Pointer & alignedAtPageSizeMask) != 0)
+                        {
+                            goto Failure;
+                        }
+                    }
+                }
+
+                totalBytes = (int)totalBytes64;
+                return true;
+
+                Failure:
+                foreach (MemoryHandle handle in pinnedBuffers)
```

Comment:
Why not let finally take care of this?

Reply:
This method unpins the buffers only either if it realizes while checking them that they are unsuitable, or if an exception happens (`Pin` may run arbitrary code).

If the method succeeds and returns `true`, it is the responsibility of `ReadScatter`/`WriteGatherAtOffsetSingleSyscallAsync` to unpin them.

Putting a try/finally at the top function (`ReadScatter`/`WriteGatherAtOffsetAsync`) would require making them `async`, and the two functions of the previous paragraph need to be either way to free the array of `long`s passed to the syscalls.

Reply:
I meant something like this here:
```
```
bool success = false;
try
{
    ...

    success = true;
    return true;
}
finally
{
    if (!success)
    {
        foreach (MemoryHandle handle in pinnedBuffers)
           ...
    }
}

pinnedBuffers = null;
totalBytes = 0;
return false;
```

`catch { cleanup; throw; }` is not ideal way to do cleanup. It unwinds the stack and makes post-mortem debugging harder.

Reply:
Thanks for the tip, I fixed it.